File: rentalcar/src/calculator.ts
import { modoCalculator } from './modo';
const EVO_DAILY_COST = 104.99;
const EVO_HOURLY_COST = 17.99;
const EVO_COST_THRESHOLD = EVO_DAILY_COST / EVO_HOURLY_COST;
const RENTAL_PER_DAY_RATE = 60;
const KM_COST = 0.1;
const TIME_API_URL = 'http://worldtimeapi.org/api/timezone/America/Vancouver';

function evoCalculator(hours: number): number {
	let totalFullDay = Math.floor(hours / 24);
	let remainingHours = hours % 24;
	if (remainingHours >= EVO_COST_THRESHOLD) {
		totalFullDay += 1;
		remainingHours = 0;
	}
	let totalCost = totalFullDay * EVO_DAILY_COST + remainingHours * EVO_HOURLY_COST;
	if (totalFullDay > 0) {
		totalCost += 1.5; // PVRT
	}
	totalCost += 1.25; // One time fee
	totalCost = totalCost * 1.12; // Tax
	return totalCost;
}

function rentalCalculator(hours: number, kms: number): number {
	let totalFullDay = Math.floor(hours / 24);
	let remainingHours = hours % 24;
	let totalDayCost = (totalFullDay + Number(Boolean(remainingHours))) * RENTAL_PER_DAY_RATE;
	let totalKmCost = kms * KM_COST;
	let totalCost = totalDayCost + totalKmCost;
	return totalCost;
}

function convertTimeStringToIsoFormat(timeString: string): string {
	// convert time in format "2022-02-02-10-00-00" to "Y-m-d H:i:sO."
	let timeArray = timeString.split('-');
	let year = timeArray[0];
	if (year.length < 4) {
		year = '20' + year;
	}
	let month = timeArray[1];
	if (month.length < 2) {
		month = '0' + month;
	}
	let day = timeArray[2];
	if (day.length < 2) {
		day = '0' + day;
	}
	let hour = timeArray[3];
	if (hour.length < 2) {
		hour = '0' + hour;
	}
	let minute = timeArray[4];
	if (minute.length < 2) {
		minute = '0' + minute;
	}
	let second = timeArray[5];
	if (second.length < 2) {
		second = '0' + second;
	}
	return `${year}-${month}-${day} ${hour}:${minute}:${second}`;
}

async function hoursCalculator(start: string, end: string): Promise<number> {
	let startTime = new Date(start);
	let endTime = new Date(end);
	let currentTime = await getCurrentTime();
	console.log(startTime);
	if (startTime > endTime) {
		return -1;
	}
	if (startTime < currentTime || endTime < currentTime) {
		console.log({ start: startTime, end: endTime, current: currentTime });
		return -2;
	}
	let difference = endTime.getTime() - startTime.getTime();
	let totalHours = difference / 1000 / 3600;
	return totalHours;
}

async function getCurrentTime(): Promise<Date> {
	try {
		let response = await fetch(TIME_API_URL);
		let data: any = await response.json(); // Add type assertion here
		let currentTime = new Date(data.datetime);
		return currentTime;
	} catch (e) {
		return new Date();
	}
}

export async function totalPriceCalculator(start: string, end: string, kms: number): Promise<any> {
	// create object to store results
	let result: any = {};
	start = convertTimeStringToIsoFormat(start);
	end = convertTimeStringToIsoFormat(end);
	let hours: number;
	try {
		hours = await hoursCalculator(start, end);
		if (hours <= 0) {
			result.status = 'Error with start and end dates. Hours cannot be negative:' + hours.toString();
			return result;
		}
		if (kms <= 0) {
			result.status = 'Error with kms. Kms cannot be negative. Kms value: ' + kms.toString();
			return result;
		}
	} catch (e: unknown) {
		return 'Error with start and end dates: ' + (e as Error).toString();
	}
	let evoTotal: string = evoCalculator(hours).toFixed(2);
	let modoTotal: string = (await modoCalculator(start, end, kms)).toFixed(2);
	let rentalTotal: string = rentalCalculator(hours, kms).toFixed(2);
	result['modo'] = modoTotal;
	result['evo'] = evoTotal;
	result['rental'] = rentalTotal;
	result['status'] = 'Success';
	return result;
}

File: Python/rentalCar/calculator.py
from rentalCar.modo import modo_calculator
import datetime
import requests
from pytz import timezone
import datetime
import requests
from pytz import timezone
from flask import request

EVO_DAILY_COST = 104.99
EVO_HOURLY_COST = 17.99
EVO_COST_THRESHOLD = EVO_DAILY_COST / EVO_HOURLY_COST
RENTAL_PER_DAY_RATE = 60
MODO_PER_DAY_RATE = 90
KM_COST = 0.10
TIME_FORMAT = "%Y-%m-%d-%H-%M-%S"
TIMEZONE = 'America/Vancouver'
TIME_API_URL = "http://worldtimeapi.org/api/timezone/America/Vancouver"


def evo_calculator(hours):
    total_full_day, remaining_hours = divmod(hours, 24)
    if remaining_hours >= EVO_COST_THRESHOLD:
        total_full_day += 1
        remaining_hours = 0
    total_cost = total_full_day * EVO_DAILY_COST + remaining_hours * EVO_HOURLY_COST
    if (total_full_day > 0):
        total_cost += 1.5  # PVRT
    total_cost += 1.25  # One time fee
    total_cost = total_cost * 1.12  # Tax
    return total_cost


def rental_calculator(hours, kms):
    total_full_day, remaining_hours = divmod(hours, 24)
    total_day_cost = (total_full_day + bool(remaining_hours)
                      ) * RENTAL_PER_DAY_RATE
    total_km_cost = kms * KM_COST
    total_cost = total_day_cost + total_km_cost
    return total_cost


def hours_calculator(start, end):
    start_time = datetime.datetime.strptime(start, TIME_FORMAT)
    end_time = datetime.datetime.strptime(end, TIME_FORMAT)
    current_time = get_current_time()
    start_time = start_time.replace(tzinfo=timezone(TIMEZONE))
    end_time = end_time.replace(tzinfo=timezone(TIMEZONE))
    if start_time > end_time:
        return -1
    if start_time < current_time or end_time < current_time:
        print({"start": start_time, "end": end_time, "current": current_time})
        return -2
    difference = end_time - start_time
    total_hours = difference.total_seconds() / 3600
    return total_hours


def get_current_time():
    try:
        response = requests.get(TIME_API_URL)
        data = response.json()
        current_time = data.get("datetime", None)
        return datetime.datetime.fromisoformat(current_time)
    except Exception as e:
        return datetime.datetime.now()


def total_price_calculator():
    start = request.args.get('start')
    end = request.args.get('end')
    kms = request.args.get('kms')
    if not all([start, end, kms]):
        return "Error with start, end, or kms"
    try:
        hours = hours_calculator(start, end)
        if hours <= 0:
            return "Error with start and end dates: "+str(hours)
        kms = float(kms)
        if kms <= 0:
            return "Error with kms"
    except Exception as e:
        return "Error with start and end dates: "+str(e)
    evo_total = str(round(evo_calculator(hours), 2))
    modo_total = str(round(modo_calculator(start, end, kms), 2))
    rental_total = str(round(rental_calculator(hours, kms), 2))
    final = ""
    final += "Evo: $"+evo_total+"<br/>"
    final += "Modo: $"+modo_total+"<br/>"
    final += "Rental: $"+rental_total+"<br/>"
    return final
